File: code/algo/getoffer/go/2/bst_find_leftbottom_val.go
package main

func findBottomLeftValue(root *TreeNode) (curVal int) {
	curHeight := 0
	var dfs func(*TreeNode, int)
	dfs = func(node *TreeNode, height int) {
		if node == nil {
			return
		}
		height++
		dfs(node.Left, height)
		dfs(node.Right, height)
		if height > curHeight {
			curHeight = height
			curVal = node.Val
		}
	}
	dfs(root, 0)
	return
}

File: code/algo/getoffer/c++/2/bst_find_leftbottom_val.cpp
#include <algorithm>

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
  int findBottomLeftValue(TreeNode *root) {
    int curVal = 0;
    int curHeight = 0;
    dfs(root, 0, curHeight, curVal);
    return curVal;
  }

private:
  void dfs(TreeNode *node, int height, int &curHeight, int &curVal) {
    if (!node) {
      return;
    }
    height++;
    dfs(node->left, height, curHeight, curVal);
    dfs(node->right, height, curHeight, curVal);
    if (height > curHeight) {
      curHeight = height;
      curVal = node->val;
    }
  }
};
